packages/rsocket-messaging/src/index.ts

```diff
@@ -0,0 +1,748 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+"use strict";
+
+import {
+  decodeCompositeMetadata,
+  decodeRoutes,
+  encodeRoutes,
+  WellKnownMimeType,
+} from "@rsocket/composite-metadata";
+import {
+  Cancellable,
+  ErrorCodes,
+  FrameTypes,
+  OnExtensionSubscriber,
+  OnNextSubscriber,
+  OnTerminalSubscriber,
+  Payload,
+  Requestable,
+  RSocket,
+  RSocketError,
+} from "@rsocket/core";
+
+export interface Codec<D> {
+  mimeType: string;
+
+  encode(entity: D): Buffer;
+  decode(buffer: Buffer): D;
+}
+
+// class AsyncFnfCallAdapter<InputDataType>
+//   implements
+//     CallAdapter<
+//       InputDataType,
+//       void,
+//       InputDataType,
+//       Promise<void> & Cancellable
+//     >
+// {
+//   _inputDataType: InputDataType;
+//   _outputDataType: void;
+
+//   createCall(
+//     rsocket: RSocket,
+//     metadata: Buffer,
+//     input: InputDataType,
+//     inputCodec: Codec<InputDataType>,
+//     _outputCodec: Codec<void> = undefined
+//   ): Promise<void> & Cancellable {
+//     return new PromiseSubscriber((s) =>
+//       rsocket.fireAndForget(
+//         {
+//           data: inputCodec.encode(input),
+//           metadata,
+//         },
+//         s
+//       )
+//     );
+//   }
+// }
+
+// class PromiseSubscriber<T>
+//   extends Promise<T>
+//   implements OnTerminalSubscriber, OnNextSubscriber, Cancellable
+// {
+//   private resolve: (value: T | PromiseLike<T>) => void;
+//   private reject: (reason?: any) => void;
+//   private cancellable: Cancellable;
+
+//   constructor(
+//     call: (subscriber: OnNextSubscriber & OnTerminalSubscriber) => Cancellable,
+//     private readonly responseCodec?: Codec<T>
+//   ) {
+//     super((resolve, reject) => {
+//       this.resolve = resolve;
+//       this.reject = reject;
+
+//       call(this);
+//     });
+//   }
+
+//   cancel(): void {
+//     this.cancellable.cancel();
+//   }
+
+//   onNext(payload: Payload, _isComplete: boolean): void {
+//     this.resolve(this.responseCodec.decode(payload.data));
+//   }
+
+//   onError(error: Error): void {
+//     this.reject(error);
+//   }
+
+//   onComplete(): void {
+//     this.resolve(undefined);
+//   }
+// }
+
+export interface RequestSpec {
+  metadata(key: string | WellKnownMimeType | number, content: Buffer): this;
+
+  request<TResponseType>(
+    exchangeFunction: (
+      rsocket: RSocket,
+      metadata: Map<string | number | WellKnownMimeType, Buffer>
+    ) => TResponseType
+  ): TResponseType;
+}
```

Comment:
Playing around with this API I found I wanted to be able to call `request()` without a provided `exchangeFunction`, or with a default exchange function that would return a standard implementation of an RSocket subscriber.

I started adding a `DefaultRespondersFactory` in my async branch for this purpose. Any objections to us including `Responder` and `Request` factories that use the internal subscriber shape?

```ts
  const subscriber = requester
    .route("RxEchoService.echoStream")
    .request();
```
or 
```ts
  const subscriber = requester
    .route("RxEchoService.echoStream")
    .request(
        DefaultRequestersFactory.requestStream(
          "Hello World",
          stringCodec,
          stringCodec,
          3
        )
    );
```

Reply:
It will not work like that for requestResponse / requestStream / channel scenarios since along with subscriber we get - we have to pass a frameHandler logic as well

Reply:
Hmm ok. I think I'm not following completely where the issue would be, so I might play around with the idea a bit and see where it falls short, but the main concern or issue right now is that essentially messaging package is only useful if the user is happy with using rxjs or async, so looking to give more options, as neither really supports backpressure in a way which the integrator can actively interface with.

Reply:
I think your idea with DefaultAdapter is a good one. However, I don't see a way to make `.request` with no args since we always have to pass data in. We can always add more adapters if it is needed